go/mysqlconn/client.go

```diff
@@ -0,0 +1,455 @@
+	return 0, fmt.Errorf("failed to interpret character set '%v'. Try using an integer value if needed", cs)
+}
+
+// clientHandshake handles the client side of the handshake.
+// Note the connection can be closed while this is running.
+func (c *Conn) clientHandshake(characterSet uint8, params *sqldb.ConnParams) error {
+	// Wait for the server initial handshake packet, and parse it.
+	data, err := c.readPacket()
+	if err != nil {
+		return sqldb.NewSQLError(CRConnHostError, "", "initial packet read failed: %v", err)
+	}
+	capabilities, cipher, err := c.parseInitialHandshakePacket(data)
+	if err != nil {
+		return sqldb.NewSQLError(CRServerHandshakeErr, SSHandshakeError, "parsing initial handshake packet failed: %v", err)
+	}
+
+	// Sanity check.
+	if capabilities&CapabilityClientProtocol41 == 0 {
+		return sqldb.NewSQLError(CRServerHandshakeErr, SSHandshakeError, "cannot connect to servers earlier than 4.1")
+	}
+
+	// If client asked for SSL, but server doesn't support it, stop right here.
+	if capabilities&CapabilityClientSSL == 0 && params.SslCert != "" && params.SslKey != "" {
+		return sqldb.NewSQLError(CRServerHandshakeErr, SSHandshakeError, "server doesn't support SSL but client asked for it")
+	}
+
+	// Remember a subset of the capabilities, so we can use them later in the protocol.
+	c.Capabilities = capabilities & (CapabilityClientDeprecateEOF)
+
+	// Build and send our handshake response 41.
+	if err := c.writeHandshakeResponse41(capabilities, cipher, characterSet, params); err != nil {
+		return err
+	}
+
+	// Read the server response.
+	response, err := c.readPacket()
+	if err != nil {
+		return fmt.Errorf("initial server response failed: %v", err)
+	}
+	switch response[0] {
+	case OKPacket:
+		// OK packet, we are authenticated. We keep going.
+	case ErrPacket:
+		return parseErrorPacket(response)
+	default:
+		// FIXME(alainjobart) handle extra auth cases and so on.
+		return fmt.Errorf("initial server response is asking for more information, not implemented yet: %v", response)
+	}
+
+	// If the server didn't support DbName in its handshake, set
+	// it now. This is what the 'mysql' client does.
+	if capabilities&CapabilityClientConnectWithDB == 0 && params.DbName != "" {
+		// Write the packet.
+		if err := c.writeComInitDB(params.DbName); err != nil {
+			return fmt.Errorf("failed to write ComInitDB packet: %v", err)
+		}
+
+		// Wait for response, should be OK.
+		response, err := c.readPacket()
+		if err != nil {
+			return fmt.Errorf("initial server response failed: %v", err)
+		}
+		switch response[0] {
+		case OKPacket:
+			// OK packet, we are authenticated.
+			return nil
+		case ErrPacket:
+			return parseErrorPacket(response)
+		default:
+			// FIXME(alainjobart) handle extra auth cases and so on.
+			return fmt.Errorf("initial server response is asking for more information, not implemented yet: %v", response)
+		}
+	}
+
+	return nil
+}
+
+func (c *Conn) parseInitialHandshakePacket(data []byte) (uint32, []byte, error) {
```

Comment:
We can consider tabularizing these functions, where input is an expected pattern, and output is the result.